File: Others/NOS/DP/5.笛卡兒積DP/lq2224.cpp
/*
 * LQ2224 [2022 国赛] 修路
 * https://www.lanqiao.cn/problems/2224/learning/
 */
#include <bits/stdc++.h>
using namespace std;
const double INF = 1e300;
#define endl '\n'

void solve() {
    int n, m, d;
    cin >> n >> m >> d;
    vector<double> A(n + 2, 0), B(m + 2, 0);
    for (int i = 1; i <= n; i++) cin >> A[i];
    for (int i = 1; i <= m; i++) cin >> B[i];
    A[n + 1] = INF;
    B[m + 1] = INF;
    sort(A.begin(), A.end());
    sort(B.begin(), B.end());
    vector<vector<array<double, 2>>> f(n + 2, vector<array<double, 2>>(m + 2, {INF, INF}));
    f[0][0][0] = 0;
    for (int i = 1; i <= n; i++) f[i][0][0] = A[i];
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            f[i][j][0] = min(f[i - 1][j][0] + A[i] - A[i - 1],
                             f[i - 1][j][1] + hypot(A[i] - B[j], d));
            f[i][j][1] = min(f[i][j - 1][1] + B[j] - B[j - 1],
                             f[i][j - 1][0] + hypot(A[i] - B[j], d));
        }
    }
    cout << fixed << setprecision(2) << min(f[n][m][0], f[n][m][1]) << endl;
    return;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    solve();
    return 0;
}

File: Others/NOS/DP/5.笛卡兒積DP/lq2224.py
"""
LQ2224 [2022 国赛] 修路
https://www.lanqiao.cn/problems/2224/learning/
"""
import math

n, m, d = map(int, input().split())
A = [0] + list(map(int, input().split())) + [float("inf")]
B = [0] + list(map(int, input().split())) + [float("inf")]
A.sort()
B.sort()

# @cache
# def f(i: int, j: int, is_A: bool) -> int:
#     if i == n and j == m:
#         return 0
#     if i > n or j > m:
#         return float("inf")
#     if is_A:
#         return min(f(i + 1, j, True) + A[i + 1] - A[i],
#                    f(i, j + 1, False) + math.hypot(A[i] - B[j + 1], d))
#     else:
#         return min(f(i, j + 1, False) + B[j + 1] - B[j],
#                    f(i + 1, j, True) + math.hypot(A[i + 1] - B[j], d))
# print(f"{f(0, 0, True):.2f}")

f = [[[float("inf")] * 2 for _ in range(m + 1)] for _ in range(n + 1)]
f[0][0][0] = 0
for i in range(1, n + 1):
    # f[i][0][0] = f[i - 1][0][0] + A[i] - A[i - 1]
    f[i][0][0] = A[i]
for i in range(1, n + 1):
    for j in range(1, m + 1):
        f[i][j][0] = min(f[i - 1][j][0] + A[i] - A[i - 1],
                         f[i - 1][j][1] + math.hypot(A[i] - B[j], d))
        f[i][j][1] = min(f[i][j - 1][1] + B[j] - B[j - 1],
                         f[i][j - 1][0] + math.hypot(A[i] - B[j], d))
print(f"{min(f[n][m][0], f[n][m][1]):.2f}")